extmod/modnetwork.c: Increase max hostname length to 32.

jimmo opened the pull request:
See also earlier PRs #12403 and #12404.

This changes from the previous limit of 15 characters. Although DHCP and mDNS allow for up to 63, ESP32 and ESP8266 only allow 32, so this seems like a reasonable limit to enforce across all ports (and avoids wasting the additional memory).

Also clarifies that `MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN` does not include the null terminator (which was unclear before).

_This work was funded through GitHub Sponsors._

jimmo committed:
extmod/modnetwork: Increase max hostname length to 32.

This changes from the previous limit of 15 characters.  Although DHCP and
mDNS allow for up to 63, ESP32 and ESP8266 only allow 32, so this seems
like a reasonable limit to enforce across all ports (and avoids wasting the
additional memory).

Also clarifies that `MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN` does not include
the null terminator (which was unclear before).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge commented on extmod/modnetwork.c:
```diff
@@ -122,7 +122,7 @@ STATIC mp_obj_t network_hostname(size_t n_args, const mp_obj_t *args) {
         size_t len;
         const char *str = mp_obj_str_get_data(args[0], &len);
-        if (len >= MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN) {
+        if (len > MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN) {
```
Would it be worth putting this block of code into a (static inline?) function, so it doesn't need to be repeated 4 times?  Like in the previous PR.

jimmo replied:
Done, in a second commit.

Rather than making a static inline, which would result in two copies per port, I just made the `if.config` version use the `network.hostname` implementation directly, which might actually be a small code size saving.

jimmo committed:
extmod/modnetwork: Forward if.config(hostname) to network.hostname.

This removes the duplicate code in cyw43, esp32, esp8266 that implements
the same logic as network.hostname.

Renames the `mod_network_hostname` (where we store the hostname value in
`.data`) to `mod_network_hostname_data` to make way for calling the shared
function `mod_network_hostname`.

And uses memcpy for mod_network_hostname_data, because the length of source
is already known and removes reliance on string data being null-terminated.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge commented on extmod/modnetwork.c:
```diff
@@ -116,20 +116,20 @@ STATIC mp_obj_t network_country(size_t n_args, const mp_obj_t *args) {
             mp_raise_ValueError(NULL);
         }
-        strcpy(mod_network_hostname, str);
+        strcpy(mod_network_hostname_data, str);
```
Would this be better as a `memcpy`, given that the length is already known, and to reduce dependence on null-terminated strings?

jimmo replied:
Updated. It's not a direct switch because you also now need to explicitly null the end.

dpgeorge replied:
It could do a memcpy with `len + 1`... that's guaranteed to work if strcpy was previously working :)

That would save a few bytes of code, and easy enough to change if/when we move away from null-terminated strings.
